### PR TITLE
[Step1-1] Bobby

### PR DESCRIPTION
**Pull request by Babywalnut**
# 안녕하세요 델마🐶
@delmaSong 

말씀해주신대로 Step 1에 대한 PR을 오늘 진행한 Task에 대해서만 나눠서 보냅니다🙈

오늘 진행한 사항에 대해서 정리해보고 느낀점 및 궁금한 점에 대해 써보겠습니다!

## 📗학습 내용
- Core ML : [문서 번역](https://babywalnut.github.io/my-blog/ios/2021/08/07/CoreML.html)

### **데이터 구조 파악**
스크린샷 2021-08-09 오후 11 08 32

주어진 데이터의 디렉터리는 test 폴더와 train 폴더로 나뉘어져 있습니다.
두 폴더에는 각각 이미지데이터를 모델에 학습시켰을 때의 결과에 부합되는 라벨명(Circle, Rectangle, Star, Triangle)의 폴더들이 존재합니다.

	test 폴더 - 학슴된 모델을 검증하는데 쓰이는 데이터

	train 폴더 - 모델을 학습시킬 때 사용되는 데이터

<br><br>
### **Create ML을 이용한 모델 학습**

figure 1
스크린샷 2021-08-09 오후 5 11 46

스크린샷 2021-08-09 오후 5 12 05

figure 2
스크린샷 2021-08-09 오후 9 06 05

스크린샷 2021-08-09 오후 9 07 09

figure 3
스크린샷 2021-08-09 오후 9 09 50

스크린샷 2021-08-09 오후 9 12 07

<!DOCTYPE html>

| |figure 1 | figure 2 | figure 3 |
| -- | -- | -- | -- |
Epoch | 25 | 100 | 50
Augmentations | Rotate | Add Noise, Blur, Crop, Expose, Flip, Rotate | Flip, Rotate
Training Accuracy | 100.0% | 25.9% | 100.0%
Validation Accuracy | 90.3% | 29.0% | 96.8%

세 번의 테스트를 통해서 figure3 세번째 시행했던 테스트의 Validation Accuracy가 가장 높았습니다. Training Accuracy에 비해 Validation Accuracy 가 낮아 Overfitting의 위험이 있다고 생각하지만 세번의 시행에서는 Validation Accuracy가 높기에 적합하다 판단하여 Create ML로 만든 모델 중에는 figure 3의 모델을 최종적으로 선택하였습니다. 

<br><br>
### **Keras를 활용한 모델 학습**

figure 1
스크린샷 2021-08-09 오후 10 40 24 

스크린샷 2021-08-09 오후 10 40 35

스크린샷 2021-08-09 오후 10 40 48

figure 2
스크린샷 2021-08-09 오후 8 03 01

스크린샷 2021-08-09 오후 8 03 24

스크린샷 2021-08-09 오후 8 02 42

<!DOCTYPE html>

| | figure 1 | figure 2 |
| -- | -- | -- |
Epoch | 100 | 200
Batch size | 10 | 8
Validation Accuracy | 60.0% | 90.0%

두번의 테스트 중에서 figure 2의 Validation Accuracy가 더 높았습니다. 처음 테스트와 비교해서 두번째 테스트에서는 Augmentation값들을 조절하여 이미지 부풀리기를 수행하였고 에폭도 100 -> 200, Batchsize 10 -> 8로 조절하여 더 많은 에폭을 돌려 전체 데이터셋을 여러번 학습하게 하였고 배치사이즈도 조절하여 더 적은 량의 데이터가 하나의 미니배치에 들어가게 하였습니다.
그결과 더 좋은 결과가 나오게되었고 Keras모델 중에는 figure 2 모델을 최종적으로 선택하였습니다.

<br><br>
### **Create ML & Keras Model Xcode Project 디렉터리에 추가**
스크린샷 2021-08-09 오후 11 32 24

선택한 두 모델을 mlmodel 확장자를 가진 파일로 프로젝트에 추가해주었습니다.

<br><br>
## 📕느낀점

 - CoLAB을 이용하여 GPU를 CoLAB서버를 통하여 사용하면 학습 시간이 비교적 빠르고 모델을 더 세밀하게 조절하여 원하는 결과를 얻을 수 있었습니다.

- Create ML을 이용하여 로컬환경에서 모델을 학습시킬 경우 Augmentation 설정을 많이 주어 데이터를 부풀리면 상당한 시간이 소요되는것 같았습니다.(다 줄 경우 거의 3시간..)

- 데이터의 양이 부족하여 데이터를 부풀린다 하더라도 새로운 데이터를 입력했을때 Overfitting이 발생하여 정확도가 떨어질 수 있을 것 같습니다.

### ~현재까지 작업한 내용은 인공지능 모델을 학습시키는 방법과 프로젝트에 추가하는 작업만 있었습니다.~ 여기까지가 Step1이었네요! ~추가한 두 모델 중 하나를 선택해 Core ML로 예측을 진행하는 작업은 코드에 대한 이해가 필요할 것 같아서 다음 작업 단위로 나눴습니다!~ (스텝3 내용이었네요😅 )작업한 내용에 대해 피드백 부탁드립니다 🙏

### CODE REVIEW

**Review by delmaSong — APPROVED** [review]
수고 많으셨습니다 바비.!
하루만에 CoreML에 대한 학습 뿐만 아니라 각각에 대한 테스트도 진행하셨군요! 속도가 놀랍네요 ㅎㅎ 
이것 자체로 스텝1 진행하신걸로 알겠습니다!
